mlir-sycl/doc/design/SYCLGridRepresentation.md

```diff
@@ -0,0 +1,244 @@
+
+### Registering operations implementing `SYCLMethodOpInterface`
+
+The `SYCLDialect::addOperations()` function is defined, overriding
+`Dialect::addOperations()`. This function calls the base implementation, but
+also **registers** operations as `SYCLMethodOpInterface` instances. In order to
+do so, a `MethodRegistry` member variable is added to the dialect. This holds a
+`(TypeID, FunctionName) -> OperationName` table so that operations can be
+registered as "methods" calling `MethodRegistry::registerMethod(mlir::TypeID,
+llvm::StringRef, llvm::StringRef)`, where the first two parameters correspond to
+the keys of the map (`getTypeID()` and `getMethodNames()`) and the last value,
+to the value of the entry (the name of the operation).
+
+This process will require no action from the users of the dialect. As this
+mechanism is implemented in the `SYCLDialect::addOperations()` function,
+defining a new function implementing `SYCLMethodOpInterface` in the TableGen
+file also generates call to register it with no further action.
+
+### Looking up operations implementing `SYCLMethodOpInterface`
+
+In the codegen side, right before generating a `sycl.call` operation, we check
+whether there is an operation implementing `SYCLMethodOpInterface` homologous to
+that function call. In order to do that, `SYCLDialect::findMethod(mlir::TypeID,
+llvm::StringRef)` can be used, passing the ID of the SYCL type the function is a
+member of and the function name. After looking up the name in its registry,
+`llvm::None` is returned if no operation can be used to replace the `sycl.call`;
+otherwise, the name of the operation is returned, which can be used to create a
+different operation instead.
+
+Note that, in case the function is a member of a base type, the `sycl.cast`
+operation used to cast to the base class is abstracted beforehand.
+
+### Canonicalization
+
+To help the analysis, canonicalization rules are introduced in such a way that
+the total number of operations is minimized, i.e., we reduce the number of
+operations by replacing combinations of operations by equivalent
+single-operations, e.g., the following code:
+
+```MLIR
+%0 = sycl.nd_item.get_global_id(%nd) {FunctionArgTypes = {!llvm.ptr<struct<...>, 4>}, FunctionName = @get_global_id, TypeName = @nd_item} : (!sycl_nd_item_1_) -> !sycl_id_1_
+%1 = sycl.id.get(%0, %idx) {FunctionArgTypes = {memref<?x!sycl_id_1_, 4>, i32}, FunctionName = @get, TypeName = @id} : (!sycl_id_1_, i32) -> i64
+```
+
+would be simplified to:
+
+```MLIR
+%1 = sycl.nd_item.get_global_id(%nd, %idx) {FunctionArgTypes = {!llvm.ptr<struct<...>, 4>, i32}, FunctionName = @get_global_id, TypeName = @nd_item} : (!sycl_nd_item_1_, i32) -> i64
+```
+
+### Function definition register
+
+After the canonicalization pass, we might obtain a module in which a new
+operation implementing `SYCLMethodOpInterface` is introduced, but no
+corresponding definition is present in the module. To avoid this situation, the
```

Comment:
The module should contain all the functions in the SYCL RT so the definition should be present in the module. Not sure I understand this point.

Reply:
Unused functions will have been removed at this point. As functions implementing this operations are not referenced in the code, they will have been removed from the module. However, they will still be in the dialect's register.

Reply:
I don't think member functions from the header would be removed because they are external functions (they are public functions) and they cannot be removed even if they aren't referenced in a module (they might be referenced by another module).  If they were removed we would not be able to generate the call to the SYCL RT and linking would fail. 

The compiler *should* generate all device functions coming from the runtime because they are parsed. IS possible one might have another module containing only an external declaration for the function, and a call to it. That module will only link if the definition of the function is found in one translation unit.

Reply:
> The compiler _should_ generate all device functions coming from the runtime because they are parsed. IS possible one might have another module containing only an external declaration for the function, and a call to it. That module will only link if the definition of the function is found in one translation unit.

As this lowering is done at late stages, we could have inlined them by this point and removed the definition or we could even have never parsed them in the first place if the function being called in the new operation's implementation is not in the original code.

Reply:
> we could even have never parsed them in the first place if the function being called in the new operation's implementation is not in the original code.

what are we doing for this case?